llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp

```diff
@@ -0,0 +1,2309 @@
+        }
+      }
+    };
+
+    if (DbgValueRanges->hasSingleLocEntry(DVA)) {
+      DbgValueDef DV = DbgValueRanges->getSingleLoc(DVA);
+      Symbol->addLocation(llvm::dwarf::DW_AT_location, /*LowPC=*/0,
+                          /*HighPC=*/-1, /*SectionOffset=*/0,
+                          /*OffsetOnEntry=*/0);
+      assert(DV.IsMemory && "Single location should be memory!");
+      AddLocation(DV);
+    } else {
+      for (DbgRangeEntry Entry : DbgValueRanges->getVariableRanges(DVA)) {
+        // These line addresses should have already been inserted into the
+        // InstrLineAddrMap, so we assume they are present here.
+        LVOffset Start = InstrLineAddrMap.at(Entry.Start.getNodePtr());
+        LVOffset End = InstrLineAddrMap.at(Entry.End.getNodePtr());
+        Symbol->addLocation(llvm::dwarf::DW_AT_location, Start, End,
+                            /*SectionOffset=*/0, /*OffsetOnEntry=*/0);
+        DbgValueDef DV = Entry.Value;
+        AddLocation(DV);
+      }
+    }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// IR Reader entry point.
+//===----------------------------------------------------------------------===//
+Error LVIRReader::createScopes() {
+  LLVM_DEBUG({
+    W.startLine() << "\n";
+    W.printString("File", getFilename());
+    W.printString("Format", FileFormatName);
+  });
+
+  // The IR Reader supports only debug records.
+  // We identify the debug input format and if it is intrinsics, it is
+  // converted to the debug records.
+  if (Error Err = LVReader::createScopes())
+    return Err;
+
+  LLVMContext Context;
+  SMDiagnostic Err;
+  std::unique_ptr<Module> M = parseIR(
+      BitCodeIR ? BitCodeIR->getMemoryBufferRef() : *TextualIR, Err, Context);
+  if (!M)
+    return createStringError(errc::invalid_argument,
+                             "Could not create IR information: %s",
+                             getFilename().str().c_str());
+
+  if (!M->getNamedMetadata("llvm.dbg.cu")) {
+    LLVM_DEBUG(dbgs() << "Skipping module without debug info\n");
+    return Error::success();
+  }
+
+  DwarfVersion = M->getDwarfVersion();
+
+  LLVM_DEBUG({ dbgs() << "\nProcess CompileUnits\n"; });
+  for (const DICompileUnit *CU : M->debug_compile_units()) {
+    LLVM_DEBUG({
+      dbgs() << "\nDICompileUnit:\n";
+      CU->dump();
+    });
+
+    // Record if the current source language is Fortran.
+    mapFortranLanguage(CU->getSourceLanguage());
+
+    CompileUnit = static_cast<LVScopeCompileUnit *>(constructElement(CU));
+    CUNode = const_cast<DICompileUnit *>(CU);
+
+    const DIFile *File = CU->getFile();
+    CompileUnit->setName(File->getFilename());
+    CompileUnit->setCompilationDirectory(File->getDirectory());
+    CompileUnit->setIsFinalized();
+
+    Root->addElement(CompileUnit);
+
+    if (options().getAttributeProducer())
+      CompileUnit->setProducer(CU->getProducer());
```

Comment:
I think we should also set the `CompileUnit` source language, i.e.
```suggestion
    if (options().getAttributeLanguage())
      CompileUnit->setSourceLanguage(LVSourceLanguage{static_cast<llvm::dwarf::SourceLanguage>(CU->getSourceLanguage())});
    if (options().getAttributeProducer())
      CompileUnit->setProducer(CU->getProducer());
```